services/wallet/market/market.go

```diff
@@ -133,9 +143,86 @@ func (pm *Manager) FetchTokenMarketValues(symbols []string, currency string) (ma
 	}
 
 	marketValues := result.(map[string]thirdparty.TokenMarketValues)
+	pm.updateMarketCache(currency, marketValues)
 	return marketValues, nil
 }
 
+func (pm *Manager) GetCachedTokenMarketValues() MarketValuesPerCurrencyAndToken {
+	pm.marketCacheLock.RLock()
+	defer pm.marketCacheLock.RUnlock()
+	return pm.marketCache
+}
+
+func (pm *Manager) GetOrFetchTokenMarketValues(symbols []string, currency string, maxAgeInSeconds int64) (map[string]thirdparty.TokenMarketValues, error) {
+	symbolsToFetchMap := make(map[string]bool)
+	symbolsToFetch := make([]string, 0, len(symbols))
+
+	now := time.Now().Unix()
+	tokenMarketValuesCache, ok := pm.GetCachedTokenMarketValues()[currency]
+	if !ok {
+		return pm.FetchTokenMarketValues(symbols, currency)
+	}
+
+	for _, symbol := range symbols {
+		marketValueSnapshot, found := tokenMarketValuesCache[symbol]
+		if !found {
+			if !symbolsToFetchMap[symbol] {
+				symbolsToFetchMap[symbol] = true
+				symbolsToFetch = append(symbolsToFetch, symbol)
+			}
+			continue
+		}
+		if now-marketValueSnapshot.UpdatedAt > maxAgeInSeconds {
+			if !symbolsToFetchMap[symbol] {
+				symbolsToFetchMap[symbol] = true
+				symbolsToFetch = append(symbolsToFetch, symbol)
+			}
+			continue
+		}
+	}
+
+	if len(symbolsToFetch) > 0 {
+		_, err := pm.FetchTokenMarketValues(symbolsToFetch, currency)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Here we're taking all the tokens that need to be requested and calling `FetchTokenMarketValues` with those token symbols.